docs/contributing/code-review.md

```diff
@@ -0,0 +1,121 @@
+This lets us treat our infrastructure as code, so we can use regular code quality practices (code review, linters, extensive documentation, etc) to keep the quality of our infrastructure high.
+However, it comes with a few caveats:
+
+1. Locally and iteratively testing terraform is impossible - as we
+   develop it, the infrastructure is deployed / modified / destroyed
+   as we go.
+2. Behavior of terraform code differs based on the current
+   state of the infrastructure (as represented in the
+   [state file](https://www.terraform.io/docs/language/state/index.html)).
+3. While `terraform plan` can tell us what terraform *thinks* it is
+   going to do, it might not always succeed. Permission errors or
+   organizational restrictions can cause it to fail. Timeouts can happen.
+   Runtime errors pop up.
+
+Thanks to all these, it is impossible to either:
+
+- Accurately gauge the impact of a terraform code change just from looking at the text
+- Automatically deploy it via continuous deployment
+
+As a result, we have some workflows that are specific to using Terraform, described below.
+
+### Changes to running infrastructure
+
+In this scenario, there is already-running infrastructure and you are deploying changes to it.
+
+For modifications to *existing infrastructure*, the suggested workflow is:
+
+
+3. 
+4. After the change has been successfully performed with `terraform apply`,
+   self-merge your PR.
+
+
+To deploy changes, follow these steps:
+
+1. **Propose changes**. Change the codebase and open a PR. **Do not deploy changes yet**. Use `terraform plan` rather than `terraform apply` do understand the change before deploying it.
+1. **Post the output of `terraform plan`**. In your PR's top comment, post the output of `terraform plan` whenever you update the PR. This helps others review the likely impact on our infrastructure.
+1. **Wait for review and approval**. Leave the PR open for other team members to review and approve.
+
+   :::{admonition} Guidelines for reviewing changes to existing infrastructure
+   :class: tip
+   Focus is on both making sure the change is minimally disruptive, and on the quality of the infrastructure design.
+   :::
+1. **Deploy the change**. Once approved, `terraform apply` your change, and make sure it works ok.
+1. **Communicate that you've deployed**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
+1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:
+   
+   1. Amend your PR with changes
+   2. Deploy with `terraform apply`
+   3. Comment that you've made a deployment
+1. **Self-merge after approval**. Once someone has approved, and you are satisfied with the resulting behavior in our infrastructure, merge the PR.
```

Comment:
I made a pass on the last four root-level bullet points. Feel free to edit my suggestion with tweaks to it as well. I wanted to make us note that we are both starting and stopping a deployment, as well as a clarification on what to do if you needed to make a changes to make it deploy successfully or not.

```suggestion
1. **Deploy the change**. Once approved, leave a comment in the PR that you will start deploying the changes. Then `terraform apply` your change and verify it works ok.
1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:

   1. Amend your PR with changes
   2. Deploy with `terraform apply`
   3. Comment that you've made a deployment
1. **Communicate that you've deployed**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
1. **Self-merge or request another review**. If you didn't need to make a change to the PR, merge the PR. If you made changes to the PR, use your judgement to decide if you should request another review or if you should self-merge it directly and just leaving a comment about the change since it was initially approved.
```